Test downsample runtime fields and security

salvatore-campagna opened the pull request:
Add YAML tests for runtime fields, document level security and field level security applied
to the downsampling operation and the downsample source and target indices.

We fail downsampling if field level security or document level security restrict access to
fields and/or documents in the source index. This is done mainly to prevent situations where
a user not allowed to read documents and/or fields on the source index is (by mistake) allowed
access to documents and/or fields in the target index, which would normally not be allowed to
have access to.

salvatore-campagna committed:
test: runtime fields with tsdb and downsampling

salvatore-campagna committed:
test: document and field level security with downsampling

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

salvatore-campagna committed:
test: fix permissions on target index

Document level and field level security is used just
to control read access to documents and fields.

salvatore-campagna commented:
I will review some security tests. I just realised that my understanding od document level and field level security was incorrect. I see they are just used to allow/prevent read access to document and fields.

salvatore-campagna committed:
test: move test to modules/lang-painless to use painless scripts

salvatore-campagna commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_tsdb_downsampling.yml:
```diff
@@ -0,0 +1,199 @@
+      reason: "downsample introduced in 8.5.0"
+
+  - do:
+      catch: '/docvalues not found for index sort field:\[\@timestamp\]/'
```
This is happening because for a `time_series` index we setup sorting by default on the `timestamp` field. As a result, at some point when setting up sorting on that field we try to get the Fielddata which is not there for a runtime field.

Should we check if the field is a runtime filed and return an error saying "Unable to sort using runtime field <field_name>"?

salvatore-campagna committed:
test: move yaml tests

salvatore-campagna committed:
test: rename indices

salvatore-campagna committed:
fix: index name

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/rest-compatibility

salvatore-campagna committed:
Merge branch 'main' into test/downsample-runtime-fields-and-security

salvatore-campagna committed:
test: fail downsampling if security is configured on the source index

salvatore-campagna committed:
fix: move variable close to if statement

salvatore-campagna commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/TransportRollupAction.java:
```diff
@@ -144,8 +149,24 @@ protected void masterOperation(
+                if (hasDocumentLevelPermissions || hasFieldLevelSecurity) {
+                    listener.onFailure(
+                        new ElasticsearchException(
+                            "Rollup forbidden for index [" + sourceIndexName + "] with document level or field level security settings."
```
I used "Rollup" here instead of "Downsample" just to be coherent with other error messages.

csoulios replied:
Aren't all messages in `TransportRollupAction` class say  "Downsample" instead of "Rollup? If not I think we should change them (in this or a follow up PR)

salvatore-campagna commented:
I included the check on security settings defined on the source index to this PR.

salvatore-campagna committed:
Update docs/changelog/90593.yaml

salvatore-campagna commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/TransportRollupAction.java:
```diff
@@ -144,8 +149,24 @@ protected void masterOperation(
+            if (indexPermissions != null) {
+                boolean hasDocumentLevelPermissions = indexPermissions.getDocumentPermissions().hasDocumentLevelPermissions();
+                boolean hasFieldLevelSecurity = indexPermissions.getFieldPermissions().hasFieldLevelSecurity();
+                if (hasDocumentLevelPermissions || hasFieldLevelSecurity) {
```
Before doing this check I also need to make sure security settings are enabled...

csoulios replied:
Is this something you plan to do in this PR or a follow up?

salvatore-campagna committed:
fix: check for document and field level security only if security is enabled

salvatore-campagna committed:
fix: check dls and fls license state

martijnvg reviewed (commented):
I left a few comments. 
Regarding the downsample with security, I think we need to test the following:

- User performing downsampling doesn't have read permissions on the source index.
- User performing downsampling doesn't have write / create_index / manage permissions on the target index.
- User has dls enabled for source index.
- User has fls enabled for source index.
- User can successfully downsample, because it has sufficient permissions on both source and target indices and no fls or dls has been configured.

I think some of this is already covered, but the test naming didn't match it.

salvatore-campagna committed:
Revert "fix: check dls and fls license state"

This reverts commit 8ee4eea131e81c39e5fd2a23bf99eff5283b58f8.

salvatore-campagna committed:
Revert "fix: check for document and field level security only if security is enabled"

This reverts commit b846647c84dea42a2520679868861fad2704b0a1.

salvatore-campagna committed:
fix: remove index user and role

salvatore-campagna committed:
fix: rename yaml test

salvatore-campagna committed:
fix: rename yaml test

salvatore-campagna committed:
fix: rewrite security tests including transport action security, dls and fls security

Here we change the security tests so to have thw following four scenarios tested:
1. Donwsample operation executed by a non-admin user
2. Downsample operation executed by an admin with field level security
3. Downsample operation executed by an admin with document level security
4. Downsample operation executed by an admin without field or document level security

csoulios commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/190_tsdb_downsampling.yml:
```diff
@@ -0,0 +1,51 @@
+---
+"Unable to define a runtime field as a time series dimension":
```
Should this test be under the `lang-painless` module? I think it should be under the tsdb test (probably `rest-api-spec/test/tsdb/70_dimension_types.yml`) or even the `rest-api-spec/test/tsdb/120_runtime_fields.yml`

csoulios commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/tsdb/20_mapping.yml:
```diff
@@ -364,3 +364,52 @@ nested fields:
                     type: keyword
+
+---
+"Unable to define a metric type for a runtime field":
```
Maybe it is worth having all tsdb/runtime fields tests in the same file `rest-api-spec/test/tsdb/120_runtime_fields.yml`

salvatore-campagna replied:
This test was there but Nik suggested to move it here. So I am not sure now where this belongs.

martijnvg replied:
Are these runtime fields yaml tests also using tsdb?

nik9000 replied:
I'm ok moving the runtime fields TSDB tests to one spot if that makes everyone feel better.

nik9000 replied:
Honestly I had them together because there weren't many. But now that there are more pulling is fine.

martijnvg replied:
imo we also can re-organise the tsdb / runtime fields tests in a followup pr.

salvatore-campagna replied:
Ok I will do that separately...moving all tests using runtime fields in a file `120_runtime_fields.yml`

csoulios commented on x-pack/plugin/rollup/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/rollup/40_runtime_fields.yml:
```diff
@@ -0,0 +1,407 @@
+          }
+  - is_true: acknowledged
+
+  # Check downsample target index runtime fields mappings
+  - do:
+      indices.get_mapping:
+        index: source
```
Here you test the `source` index and not the `target` index. Is that right? Comment says "check downsample target index"

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/TransportRollupAction.java:
```diff
@@ -144,8 +149,24 @@ protected void masterOperation(
+            if (indexPermissions != null) {
+                boolean hasDocumentLevelPermissions = indexPermissions.getDocumentPermissions().hasDocumentLevelPermissions();
+                boolean hasFieldLevelSecurity = indexPermissions.getFieldPermissions().hasFieldLevelSecurity();
+                if (hasDocumentLevelPermissions || hasFieldLevelSecurity) {
```
I thought we had agreed that if security is enabled on the index, downsampling would be entirely not allowed. 

Here, I see it is disabled only if DLS or FLS is enabled. This may make sense, but I wanted to make sure we are all onboard with this decision 

cc @nik9000 @martijnvg

salvatore-campagna replied:
I discussed this with @martijnvg , there are two levels of security here:
1. security associated to the user/role that hits the API endpoint: this is the first level of security. `TransportRollupAction` uses "indices:admin/xpack/downsample" which means that calling downsample API requires `admin` privileges
2. if the API call is allowed then we check if DLS or FLS is defined.

This code is checking just (2). Responsibility of (1) does not belong to "application logic" but is a cross cutting concern that is delegated to security interceptors.

See test named "Downsample user missing admin permissions to run the downsample operation" which tests (1) and the other 3 tests which check the other 3 cases:
1. Admin with DLS => fails because of DLS
2. Admin with FLS => fials because of FLS
3. Admin without DLS and without FLS => succeeds because it has admin rights and no DLS or FLS defined

salvatore-campagna replied:
@csoulios is there any reason why this was implemented requiring admin permissions to run the downsample operation? Should't be better to have a specific set of permissions (something like `downsample` role) which allows running downsampling if the user has:
1. read source index permissions
2. create and write target index permissions
3. any other permission required to do things like copy settings/mappings/...

csoulios replied:
It was first implemented with the strictest permissions. What you write sounds reasonable to me

csoulios reviewed (commented):
Very good work and with a comprehensive test suite. 

I have left some small comments.

salvatore-campagna committed:
fix: check runtime mappings in the target index

salvatore-campagna committed:
test: move @timestamp runtime field test

salvatore-campagna committed:
test: make the test consistent with othr tests

salvatore-campagna commented:
@elasticsearchmachine run elasticsearch-ci/bwc

salvatore-campagna committed:
test: move test to 70_dimension_types

salvatore-campagna committed:
test: move test to 70_dimension_types

salvatore-campagna committed:
Merge branch 'main' into test/downsample-runtime-fields-and-security

martijnvg commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/security/authz/80_downsample.yml:
```diff
@@ -0,0 +1,263 @@
+          {
+            "indices": [
+              {
+                "names": ["source", "target"],
```
I think downsample would still work with just `all` privilege on the `source` index?

salvatore-campagna replied:
Actually not...a test fails with:
`action [indices:data/read/search] is unauthorized for user [downsample-user] with effective roles [downsample-role] on indices [target], this action is granted by the index privileges [read,all]`

salvatore-campagna committed:
fix: remove target index

salvatore-campagna committed:
test: remove search api call which needs target read permissions

salvatore-campagna committed:
test: downsampling using multiple indices

martijnvg reviewed (approved):
I left a few comments, otherwise LGTM.

salvatore-campagna committed:
fix: remove target index

salvatore-campagna commented:
We fail downsampling if field level security or document level security
restrict access to fields and/or documents in the source index.
This is done mainly to prevent situations where a user not allowed to
read documents and/or fields on the source index is (by mistake) allowed
access to documents and/or fields in the target index, which would normally
not be allowed access to.
